pkg/controller/deployment/deployment_controller.go

```diff
@@ -557,9 +597,23 @@ func (dc *DeploymentController) getNewRC(deployment extensions.Deployment) (*api
 		dc.rcExpectations.DeleteExpectations(dKey)
 		return nil, fmt.Errorf("error creating replication controller: %v", err)
 	}
+
+	if err = dc.updateDeploymentVersion(deployment, newVersion); err != nil {
+		return createdRC, err
+	}
+
 	return createdRC, nil
 }
 
+func (dc *DeploymentController) updateDeploymentVersion(deployment extensions.Deployment, version string) error {
+	if deployment.Annotations == nil {
+		deployment.Annotations = make(map[string]string)
+	}
+	deployment.Annotations[deploymentutil.VersionAnnotation] = version
```

Comment:
Deployment version is from observing current RCs but not from this deployment version annotation. We look at the deployment's RCs, get the max version number among them, determine what version number we should give to the new RC, and update deployment version number as a reference. 

Even if we implement and use deployment generation numbers, we cannot count on it; otherwise, when the deployment is killed and then re-created (generation number becomes 1) we lost the actual version number.